File: src/lib/permissions.ts
/**
 * K-LENS RBAC Permission System
 * Industrial-grade Role-Based Access Control for Frontend
 */

// ===================== TYPES =====================

export type Role = 'ADMIN' | 'MANAGER' | 'ENGINEER' | 'SAFETY_OFFICER' | 'OPERATOR';

export type Permission = 
  // User Management
  | 'USER_CREATE' | 'USER_DELETE' | 'USER_EDIT' | 'USER_VIEW_ALL' 
  | 'USER_ASSIGN_SHIFT' | 'USER_RESET_MFA'
  // Documents
  | 'DOC_UPLOAD' | 'DOC_EDIT' | 'DOC_DELETE' | 'DOC_APPROVE_DELETE'
  | 'DOC_VERSION' | 'DOC_VIEW_ALL' | 'DOC_VIEW_DEPT' | 'DOC_VIEW_ASSIGNED'
  | 'DOC_TAG_RISK' | 'DOC_AUDIT_VIEW'
  // Knowledge Graph
  | 'GRAPH_CONFIG' | 'GRAPH_VIEW_ALL' | 'GRAPH_VIEW_DEPT' 
  | 'GRAPH_VIEW_RISK' | 'GRAPH_VIEW_LOCAL'
  // IoT & UNS
  | 'IOT_CONFIG' | 'IOT_VIEW_ALL' | 'IOT_CALIBRATE' 
  | 'IOT_ALERTS_VIEW' | 'IOT_MONITOR'
  // Emergency & Override
  | 'EMERGENCY_GRANT' | 'EMERGENCY_REQUEST' | 'EMERGENCY_AUDIT'
  // Compliance
  | 'COMPLIANCE_VIEW' | 'COMPLIANCE_WRITE' | 'COMPLIANCE_FLAG'
  // Shift Management
  | 'SHIFT_BYPASS' | 'SHIFT_VIEW_ALL' | 'SHIFT_MANAGE'
  // Admin
  | 'ADMIN_SETTINGS' | 'ADMIN_GHOST_VIEW' | 'ADMIN_TRACE' | 'ADMIN_APPROVE';

// ===================== ROLE-PERMISSION MAPPING =====================

const ROLE_PERMISSIONS: Record<Role, Set<Permission>> = {
  ADMIN: new Set([
    // Admin has ALL permissions
    'USER_CREATE', 'USER_DELETE', 'USER_EDIT', 'USER_VIEW_ALL', 'USER_ASSIGN_SHIFT', 'USER_RESET_MFA',
    'DOC_UPLOAD', 'DOC_EDIT', 'DOC_DELETE', 'DOC_APPROVE_DELETE', 'DOC_VERSION', 'DOC_VIEW_ALL', 
    'DOC_VIEW_DEPT', 'DOC_VIEW_ASSIGNED', 'DOC_TAG_RISK', 'DOC_AUDIT_VIEW',
    'GRAPH_CONFIG', 'GRAPH_VIEW_ALL', 'GRAPH_VIEW_DEPT', 'GRAPH_VIEW_RISK', 'GRAPH_VIEW_LOCAL',
    'IOT_CONFIG', 'IOT_VIEW_ALL', 'IOT_CALIBRATE', 'IOT_ALERTS_VIEW', 'IOT_MONITOR',
    'EMERGENCY_GRANT', 'EMERGENCY_REQUEST', 'EMERGENCY_AUDIT',
    'COMPLIANCE_VIEW', 'COMPLIANCE_WRITE', 'COMPLIANCE_FLAG',
    'SHIFT_BYPASS', 'SHIFT_VIEW_ALL', 'SHIFT_MANAGE',
    'ADMIN_SETTINGS', 'ADMIN_GHOST_VIEW', 'ADMIN_TRACE', 'ADMIN_APPROVE',
  ]),
  
  MANAGER: new Set([
    'USER_VIEW_ALL', 'USER_ASSIGN_SHIFT',
    'DOC_VIEW_ALL', 'DOC_APPROVE_DELETE', 'DOC_AUDIT_VIEW',
    'GRAPH_VIEW_ALL',
    'IOT_VIEW_ALL',
    'EMERGENCY_REQUEST',
    'COMPLIANCE_VIEW',
    'SHIFT_BYPASS', 'SHIFT_VIEW_ALL', 'SHIFT_MANAGE',
    'ADMIN_TRACE', 'ADMIN_APPROVE',
  ]),
  
  ENGINEER: new Set([
    'DOC_UPLOAD', 'DOC_EDIT', 'DOC_VERSION', 'DOC_VIEW_DEPT', 'DOC_TAG_RISK',
    'GRAPH_VIEW_DEPT',
    'IOT_CALIBRATE', 'IOT_VIEW_ALL',
    'EMERGENCY_REQUEST',
    'COMPLIANCE_VIEW',
    'SHIFT_BYPASS',
  ]),
  
  SAFETY_OFFICER: new Set([
    'DOC_VIEW_ALL', 'DOC_AUDIT_VIEW',
    'GRAPH_VIEW_RISK', 'GRAPH_VIEW_ALL',
    'IOT_ALERTS_VIEW', 'IOT_VIEW_ALL',
    'EMERGENCY_AUDIT',
    'COMPLIANCE_VIEW', 'COMPLIANCE_WRITE', 'COMPLIANCE_FLAG',
    'SHIFT_BYPASS', 'SHIFT_VIEW_ALL',
  ]),
  
  OPERATOR: new Set([
    'DOC_VIEW_ASSIGNED',
    'GRAPH_VIEW_LOCAL',
    'IOT_MONITOR',
    'EMERGENCY_REQUEST',
    'COMPLIANCE_VIEW',
  ]),
};

// ===================== HELPER FUNCTIONS =====================

/**
 * Check if a role has a specific permission
 */
export function hasPermission(role: string | undefined, permission: Permission): boolean {
  if (!role) return false;
  const normalizedRole = role.toUpperCase() as Role;
  const rolePerms = ROLE_PERMISSIONS[normalizedRole];
  if (!rolePerms) return false;
  return rolePerms.has(permission);
}

/**
 * Check if a role has ANY of the specified permissions
 */
export function hasAnyPermission(role: string | undefined, permissions: Permission[]): boolean {
  return permissions.some(p => hasPermission(role, p));
}

/**
 * Check if a role has ALL of the specified permissions
 */
export function hasAllPermissions(role: string | undefined, permissions: Permission[]): boolean {
  return permissions.every(p => hasPermission(role, p));
}

/**
 * Get all permissions for a role
 */
export function getRolePermissions(role: string | undefined): Permission[] {
  if (!role) return [];
  const normalizedRole = role.toUpperCase() as Role;
  const rolePerms = ROLE_PERMISSIONS[normalizedRole];
  if (!rolePerms) return [];
  return Array.from(rolePerms);
}

/**
 * Check if role can bypass shift restrictions
 */
export function canBypassShift(role: string | undefined): boolean {
  return hasPermission(role, 'SHIFT_BYPASS');
}

// ===================== ROLE METADATA =====================

export const ROLE_METADATA: Record<Role, { 
  displayName: string; 
  description: string; 
  color: string;
  icon: string;
}> = {
  ADMIN: {
    displayName: 'System Administrator',
    description: 'Full technical control. Can manage users, reset MFA, view ghost accounts.',
    color: 'amber',
    icon: 'Crown',
  },
  MANAGER: {
    displayName: 'Plant Manager',
    description: 'Full operational control. Can approve deletions, trace activity, manage shifts.',
    color: 'blue',
    icon: 'Shield',
  },
  ENGINEER: {
    displayName: 'Senior Engineer',
    description: 'Technical specialist. Can upload documents, calibrate sensors, tag risks.',
    color: 'purple',
    icon: 'Wrench',
  },
  SAFETY_OFFICER: {
    displayName: 'Safety Officer',
    description: 'Compliance guard. Global read access, can flag violations and audit.',
    color: 'emerald',
    icon: 'ShieldCheck',
  },
  OPERATOR: {
    displayName: 'Operator',
    description: 'Frontline worker. Shift-locked, view assigned documents and local systems.',
    color: 'slate',
    icon: 'User',
  },
};

// ===================== PERMISSION CATEGORIES =====================

export const PERMISSION_CATEGORIES = {
  USER_MGMT: 'User Management',
  DOCUMENTS: 'Documents',
  KNOWLEDGE: 'Knowledge Graph',
  IOT: 'IoT & UNS',
  EMERGENCY: 'Emergency',
  COMPLIANCE: 'Compliance',
  SHIFT: 'Shift Management',
  ADMIN: 'Administration',
} as const;

export const PERMISSION_METADATA: Record<Permission, { 
  description: string; 
  category: keyof typeof PERMISSION_CATEGORIES;
  riskLevel: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
}> = {
  // User Management
  USER_CREATE: { description: 'Create new users', category: 'USER_MGMT', riskLevel: 'HIGH' },
  USER_DELETE: { description: 'Delete/Archive users', category: 'USER_MGMT', riskLevel: 'CRITICAL' },
  USER_EDIT: { description: 'Edit user profiles', category: 'USER_MGMT', riskLevel: 'MEDIUM' },
  USER_VIEW_ALL: { description: 'View all users', category: 'USER_MGMT', riskLevel: 'LOW' },
  USER_ASSIGN_SHIFT: { description: 'Assign shifts to users', category: 'USER_MGMT', riskLevel: 'MEDIUM' },
  USER_RESET_MFA: { description: 'Reset MFA for users', category: 'USER_MGMT', riskLevel: 'HIGH' },
  
  // Documents
  DOC_UPLOAD: { description: 'Upload documents', category: 'DOCUMENTS', riskLevel: 'MEDIUM' },
  DOC_EDIT: { description: 'Edit documents', category: 'DOCUMENTS', riskLevel: 'MEDIUM' },
  DOC_DELETE: { description: 'Delete documents', category: 'DOCUMENTS', riskLevel: 'HIGH' },
  DOC_APPROVE_DELETE: { description: 'Approve document deletions', category: 'DOCUMENTS', riskLevel: 'CRITICAL' },
  DOC_VERSION: { description: 'Manage document versions', category: 'DOCUMENTS', riskLevel: 'MEDIUM' },
  DOC_VIEW_ALL: { description: 'View all documents', category: 'DOCUMENTS', riskLevel: 'LOW' },
  DOC_VIEW_DEPT: { description: 'View department documents', category: 'DOCUMENTS', riskLevel: 'LOW' },
  DOC_VIEW_ASSIGNED: { description: 'View assigned documents only', category: 'DOCUMENTS', riskLevel: 'LOW' },
  DOC_TAG_RISK: { description: 'Tag documents as High Risk', category: 'DOCUMENTS', riskLevel: 'HIGH' },
  DOC_AUDIT_VIEW: { description: 'View document audit history', category: 'DOCUMENTS', riskLevel: 'MEDIUM' },
  
  // Knowledge Graph
  GRAPH_CONFIG: { description: 'Configure graph schema', category: 'KNOWLEDGE', riskLevel: 'HIGH' },
  GRAPH_VIEW_ALL: { description: 'View full knowledge graph', category: 'KNOWLEDGE', riskLevel: 'LOW' },
  GRAPH_VIEW_DEPT: { description: 'View department graph nodes', category: 'KNOWLEDGE', riskLevel: 'LOW' },
  GRAPH_VIEW_RISK: { description: 'View risk-related nodes', category: 'KNOWLEDGE', riskLevel: 'MEDIUM' },
  GRAPH_VIEW_LOCAL: { description: 'View local unit nodes', category: 'KNOWLEDGE', riskLevel: 'LOW' },
  
  // IoT
  IOT_CONFIG: { description: 'Configure sensors', category: 'IOT', riskLevel: 'HIGH' },
  IOT_VIEW_ALL: { description: 'View all IoT dashboards', category: 'IOT', riskLevel: 'LOW' },
  IOT_CALIBRATE: { description: 'Calibrate sensors', category: 'IOT', riskLevel: 'MEDIUM' },
  IOT_ALERTS_VIEW: { description: 'View safety alerts', category: 'IOT', riskLevel: 'LOW' },
  IOT_MONITOR: { description: 'Monitor dashboards', category: 'IOT', riskLevel: 'LOW' },
  
  // Emergency
  EMERGENCY_GRANT: { description: 'Grant emergency override', category: 'EMERGENCY', riskLevel: 'CRITICAL' },
  EMERGENCY_REQUEST: { description: 'Request emergency override', category: 'EMERGENCY', riskLevel: 'HIGH' },
  EMERGENCY_AUDIT: { description: 'Audit override history', category: 'EMERGENCY', riskLevel: 'MEDIUM' },
  
  // Compliance
  COMPLIANCE_VIEW: { description: 'View compliance reports', category: 'COMPLIANCE', riskLevel: 'LOW' },
  COMPLIANCE_WRITE: { description: 'Write compliance reports', category: 'COMPLIANCE', riskLevel: 'MEDIUM' },
  COMPLIANCE_FLAG: { description: 'Flag safety violations', category: 'COMPLIANCE', riskLevel: 'HIGH' },
  
  // Shift
  SHIFT_BYPASS: { description: 'Bypass shift restrictions', category: 'SHIFT', riskLevel: 'MEDIUM' },
  SHIFT_VIEW_ALL: { description: 'View all shift schedules', category: 'SHIFT', riskLevel: 'LOW' },
  SHIFT_MANAGE: { description: 'Manage shift schedules', category: 'SHIFT', riskLevel: 'MEDIUM' },
  
  // Admin
  ADMIN_SETTINGS: { description: 'Access admin settings', category: 'ADMIN', riskLevel: 'HIGH' },
  ADMIN_GHOST_VIEW: { description: 'View ghost/archived users', category: 'ADMIN', riskLevel: 'MEDIUM' },
  ADMIN_TRACE: { description: 'Trace user activity', category: 'ADMIN', riskLevel: 'HIGH' },
  ADMIN_APPROVE: { description: 'Approve nuclear key requests', category: 'ADMIN', riskLevel: 'CRITICAL' },
};

File: backend-python/app/core/permissions.py
"""
K-LENS RBAC Permission System
Industrial-grade Role-Based Access Control with scope validation
"""

from enum import Enum
from typing import List, Optional, Set
from functools import lru_cache


class Role(str, Enum):
    """Industrial role hierarchy"""
    ADMIN = "ADMIN"           # IT / System Admin - Full technical control
    MANAGER = "MANAGER"       # Plant Manager - Full operational control
    ENGINEER = "ENGINEER"     # Senior Engineer - Technical specialist
    SAFETY_OFFICER = "SAFETY_OFFICER"  # Compliance Guard - Global read, compliance write
    OPERATOR = "OPERATOR"     # Frontline Worker - Shift-locked, restricted access


class Permission(str, Enum):
    """All available permissions in K-LENS"""
    # User Management
    USER_CREATE = "USER_CREATE"
    USER_DELETE = "USER_DELETE"
    USER_EDIT = "USER_EDIT"
    USER_VIEW_ALL = "USER_VIEW_ALL"
    USER_ASSIGN_SHIFT = "USER_ASSIGN_SHIFT"
    USER_RESET_MFA = "USER_RESET_MFA"
    
    # Documents
    DOC_UPLOAD = "DOC_UPLOAD"
    DOC_EDIT = "DOC_EDIT"
    DOC_DELETE = "DOC_DELETE"
    DOC_APPROVE_DELETE = "DOC_APPROVE_DELETE"
    DOC_VERSION = "DOC_VERSION"
    DOC_VIEW_ALL = "DOC_VIEW_ALL"
    DOC_VIEW_DEPT = "DOC_VIEW_DEPT"
    DOC_VIEW_ASSIGNED = "DOC_VIEW_ASSIGNED"
    DOC_TAG_RISK = "DOC_TAG_RISK"
    DOC_AUDIT_VIEW = "DOC_AUDIT_VIEW"
    
    # Knowledge Graph
    GRAPH_CONFIG = "GRAPH_CONFIG"
    GRAPH_VIEW_ALL = "GRAPH_VIEW_ALL"
    GRAPH_VIEW_DEPT = "GRAPH_VIEW_DEPT"
    GRAPH_VIEW_RISK = "GRAPH_VIEW_RISK"
    GRAPH_VIEW_LOCAL = "GRAPH_VIEW_LOCAL"
    
    # IoT & UNS
    IOT_CONFIG = "IOT_CONFIG"
    IOT_VIEW_ALL = "IOT_VIEW_ALL"
    IOT_CALIBRATE = "IOT_CALIBRATE"
    IOT_ALERTS_VIEW = "IOT_ALERTS_VIEW"
    IOT_MONITOR = "IOT_MONITOR"
    
    # Emergency & Override
    EMERGENCY_GRANT = "EMERGENCY_GRANT"
    EMERGENCY_REQUEST = "EMERGENCY_REQUEST"
    EMERGENCY_AUDIT = "EMERGENCY_AUDIT"
    
    # Compliance
    COMPLIANCE_VIEW = "COMPLIANCE_VIEW"
    COMPLIANCE_WRITE = "COMPLIANCE_WRITE"
    COMPLIANCE_FLAG = "COMPLIANCE_FLAG"
    
    # Shift Management
    SHIFT_BYPASS = "SHIFT_BYPASS"
    SHIFT_VIEW_ALL = "SHIFT_VIEW_ALL"
    SHIFT_MANAGE = "SHIFT_MANAGE"
    
    # Admin
    ADMIN_SETTINGS = "ADMIN_SETTINGS"
    ADMIN_GHOST_VIEW = "ADMIN_GHOST_VIEW"
    ADMIN_TRACE = "ADMIN_TRACE"
    ADMIN_APPROVE = "ADMIN_APPROVE"


# Static role-permission mapping (in-memory for fast lookups)
ROLE_PERMISSIONS: dict[Role, Set[Permission]] = {
    Role.ADMIN: set(Permission),  # Admin has ALL permissions
    
    Role.MANAGER: {
        Permission.USER_VIEW_ALL,
        Permission.USER_ASSIGN_SHIFT,
        Permission.DOC_VIEW_ALL,
        Permission.DOC_APPROVE_DELETE,
        Permission.DOC_AUDIT_VIEW,
        Permission.GRAPH_VIEW_ALL,
        Permission.IOT_VIEW_ALL,
        Permission.EMERGENCY_REQUEST,
        Permission.COMPLIANCE_VIEW,
        Permission.SHIFT_BYPASS,
        Permission.SHIFT_VIEW_ALL,
        Permission.SHIFT_MANAGE,
        Permission.ADMIN_TRACE,
        Permission.ADMIN_APPROVE,
    },
    
    Role.ENGINEER: {
        Permission.DOC_UPLOAD,
        Permission.DOC_EDIT,
        Permission.DOC_VERSION,
        Permission.DOC_VIEW_DEPT,
        Permission.DOC_TAG_RISK,
        Permission.GRAPH_VIEW_DEPT,
        Permission.IOT_CALIBRATE,
        Permission.IOT_VIEW_ALL,
        Permission.EMERGENCY_REQUEST,
        Permission.COMPLIANCE_VIEW,
        Permission.SHIFT_BYPASS,
    },
    
    Role.SAFETY_OFFICER: {
        Permission.DOC_VIEW_ALL,
        Permission.DOC_AUDIT_VIEW,
        Permission.GRAPH_VIEW_RISK,
        Permission.GRAPH_VIEW_ALL,
        Permission.IOT_ALERTS_VIEW,
        Permission.IOT_VIEW_ALL,
        Permission.EMERGENCY_AUDIT,
        Permission.COMPLIANCE_VIEW,
        Permission.COMPLIANCE_WRITE,
        Permission.COMPLIANCE_FLAG,
        Permission.SHIFT_BYPASS,
        Permission.SHIFT_VIEW_ALL,
    },
    
    Role.OPERATOR: {
        Permission.DOC_VIEW_ASSIGNED,
        Permission.GRAPH_VIEW_LOCAL,
        Permission.IOT_MONITOR,
        Permission.EMERGENCY_REQUEST,
        Permission.COMPLIANCE_VIEW,
    },
}


def has_permission(role: str, permission: str | Permission) -> bool:
    """
    Check if a role has a specific permission.
    
    Args:
        role: User's role string
        permission: Permission to check (string or enum)
        
    Returns:
        True if role has permission, False otherwise
    """
    try:
        role_enum = Role((role or "").upper())
    except ValueError:
        return False
    
    if isinstance(permission, str):
        try:
            permission = Permission(permission)
        except ValueError:
            return False
    
    role_perms = ROLE_PERMISSIONS.get(role_enum, set())
    return permission in role_perms


def get_role_permissions(role: str) -> List[str]:
    """
    Get all permissions for a role.
    
    Args:
        role: User's role string
        
    Returns:
        List of permission strings
    """
    try:
        role_enum = Role((role or "").upper())
    except ValueError:
        return []
    
    role_perms = ROLE_PERMISSIONS.get(role_enum, set())
    return [p.value for p in role_perms]


def can_bypass_shift(role: str) -> bool:
    """
    Check if a role can bypass shift restrictions.
    
    Args:
        role: User's role string
        
    Returns:
        True if role can bypass shift, False otherwise
    """
    return has_permission(role, Permission.SHIFT_BYPASS)


def get_role_display_name(role: str) -> str:
    """Get human-readable role name"""
    display_names = {
        "ADMIN": "System Administrator",
        "MANAGER": "Plant Manager",
        "ENGINEER": "Senior Engineer",
        "SAFETY_OFFICER": "Safety Officer",
        "OPERATOR": "Operator",
    }
    return display_names.get(role, role)


def get_role_description(role: str) -> str:
    """Get role description"""
    descriptions = {
        "ADMIN": "Full technical control. Can manage users, reset MFA, view ghost accounts.",
        "MANAGER": "Full operational control. Can approve deletions, trace activity, manage shifts.",
        "ENGINEER": "Technical specialist. Can upload documents, calibrate sensors, tag risks.",
        "SAFETY_OFFICER": "Compliance guard. Global read access, can flag violations and audit.",
        "OPERATOR": "Frontline worker. Shift-locked, view assigned documents and local systems.",
    }
    return descriptions.get(role, "")


# Permission category groups for UI
PERMISSION_CATEGORIES = {
    "USER_MGMT": "User Management",
    "DOCUMENTS": "Documents",
    "KNOWLEDGE": "Knowledge Graph",
    "IOT": "IoT & UNS",
    "EMERGENCY": "Emergency",
    "COMPLIANCE": "Compliance",
    "SHIFT": "Shift Management",
    "ADMIN": "Administration",
}
